[instance] single instance optimization feature

abtink opened the pull request:
This commit adds single-instance optimization feature. This feature is enabled through configure option `--enable-single-instance-opt` and by default it is disabled.

This feature aims to simplify the code and also reduce memory/RAM use of OpenThread by implementing two optimizations:

(1) OpenThread objects/classes typically keep a reference to a higher-level object (e.g., many classes keep track of `ThreadNetif` reference which is expected as an input parameter of the object's constructor). This is used by the object to access other objects/methods within OpenThread class hierarchy including the containing `otInstance`. When single-instance optimization is enabled the reference member variables are removed and instead global functions are used to access the singleton objects from one class to the other.

To implement this, a group of `<Object>Tracker` classes are defined (e.g., `ThreadNetifTracker`, `Ip6Tracker, etc.) which keep track of an OpenThread object reference. The tracker classes are used as base class of other OpenThread classes which require tracking of a reference to another object.

The `Tracker` classes also provide implementation of `GetInstance()` method which returns a pointer to containing `otInstance` strcut. This is then inherited by all sub-classes and thus removing the need to
define this in every other class.

(2) OpenThread objects which provide a callback/handler (e.g., `Timer` ,`Tasklet`, etc.) have `void *mContext` member variable which is used to keep track of the owner of the object. When single-instance optimization is enabled the `mConext` member variable is removed since the owner is expected to be a singleton and can be uniquely determined from the callback function.

To implement this, two changes are made: First the handler methods are modified to provide a reference to the object (e.g., `Timer` handler will provide a `Timer &aTimer` as a parameter of its handler callback). Second, a new base class `Context` is introduced which hides the implementation providing an arbitrary context information and classes that provide handler callback publicly inherit the `Context`. A new static method `GetOwner(aContext)` is added to classes which own a callback providing object. This method help convert a `Context` to the reference to the owner class object.

abtink commented on configure.ac:
```diff
@@ -507,6 +507,37 @@ AC_MSG_CHECKING([sould NCP support UART])
+# Single Instance Optimization
+#
+
+AC_ARG_ENABLE(single-instance-opt,
```
We already have `OPENTHREAD_MULTIPLE_INSTANCE` as an option. It's mainly used for Windows platform and we check it with `#ifdef` instead of `#if` (would be good to change this to follow `#if` model later and also add it as a `configure.ac` option).

Here is how I see this:
 - `OPENTHREAD_MULTIPLE_INSTANCE` (if defined) determines if we have multiple ot instance vs a single instance (default case we don't have this defined, so we are single-instance). 
- `OPENTHREAD_ENABLE_SINGLE_INSTANCE_OPTIMIZATION` is used to enable the optimization feature.

With this model we can have single-instance case with or without the optimization enabled. 

Thoughts?

jwhui replied:
What is the use case for having a single instance but *without* the single instance optimization?  In the end, the more configuration options we have, the more configuration permutations we have to test.

abtink replied:
Good point! I don't think we have a current/immediate use-case for this. The only thing that may be easier with optimization disabled are some of the unit tests (e.g., timer unit test). But I think there should be ways to make it work.

I am fine with combining the options into one: 

- `OPENTHREAD_ENABLE_MULTIPLE_INSTANCES` (controlled with `--enable-multiple-instances`): If enabled, we support multiple instances, otherwise we support single instance with optimization always enabled in single instance case.

jwhui replied:
Yes, I prefer that approach 👍

abtink commented:
Changes in latest puhsed commit:
- Added `--enable-multiple-instances` and its corresponding `OPENTHREAD_ENABLE_MULTIPLE_INSTANCES` option (default is disabled). In single instance mode optimizations are always included.
- Updated the unit tests to ensure they work ok with single and/or multiple instance cases.

jwhui commented on src/core/openthread-single-instance.h:
```diff
@@ -0,0 +1,107 @@
+ot::MeshForwarder &otGetMeshForwarder(void);
+
+/**
+ * This function returns a reference to the single TimerShceduler instance.
```
typo: `TimerShceduler` -> `TimerScheduler`

jwhui commented on src/core/common/context.hpp:
```diff
@@ -0,0 +1,102 @@
+
+/**
+ * @file
+ *   This file includes definitions for a arbitrary context information.
```
Suggestion:
> This file includes definitions for maintaining a pointer to arbitrary context information.

jwhui commented on src/core/common/context.hpp:
```diff
@@ -0,0 +1,102 @@
+ * @addtogroup core-context
+ *
+ * @brief
+ *   This module includes definitions for arbitrary context information.
```
See comment above.

jwhui commented on src/core/common/context.hpp:
```diff
@@ -0,0 +1,102 @@
+ */
+
+/**
+ * This class implements an arbitrary (callback) context information.
```
See comment above.

jwhui commented on src/core/common/tracker.hpp:
```diff
@@ -0,0 +1,355 @@
+ *
+ */
+template <typename TrackedType>
+class Tracker
```
nit: "Tracker" seems to imply that the object is moving around.  Would "Locator" be more appropriate here?

abtink replied:
Yes. Thanks. I like the "Locator" option (e.g., `ThreadNetifLocator` `InstanceLocator`...

(I went back and forth between "Tracker" and "Keeper"...also considered "Maintainer", "Saver", and simply calling it "Base"... :)...)

jwhui commented on src/core/common/tracker.hpp:
```diff
@@ -0,0 +1,355 @@
+ * This class implements tracker for  otInstance object
+ *
+ */
+
```
Extra newline.

jwhui commented on src/core/meshcop/dataset.hpp:
```diff
@@ -193,7 +194,7 @@ class Dataset
     uint8_t    mTlvs[kMaxSize];  ///< The Dataset buffer
     uint16_t   mLength;          ///< The number of valid bytes in @var mTlvs
-    otInstance *mInstance;       ///< The pointer to an OpenThread instance
+
```
Extra newline?

jwhui commented:
Thanks for changing the name from Track* to Locator*.  One nit is that the alphabetical ordering also needs to be resolved with the name change.

abtink commented:
Yes! Indeed...Will take care of this (in Makefiles and order of header file includes)...  :)

abtink commented:
Fixed the orders in the new pushed commit....

jwhui commented on etc/visual-studio/libopenthread.vcxproj.filters:
```diff
@@ -335,12 +338,18 @@
     <ClInclude Include="..\..\src\core\common\code_utils.hpp">
       <Filter>Header Files\common</Filter>
     </ClInclude>
+    <ClInclude Include="..\..\src\core\common\cotext.hpp">
```
typo: `cotext` -> `context`

abtink replied:
Nice catch! Interesting that builds were passing with this.

jwhui commented on src/core/common/locator.hpp:
```diff
@@ -0,0 +1,341 @@
+    /**
+     * This method returns the pointer to the parent otInstance structure.
+     *
+     * @returns The pointer to the parent otInstance structure.
```
Document that this can return `NULL`.

abtink replied:
Will do...

jwhui commented on src/core/common/locator.hpp:
```diff
@@ -0,0 +1,341 @@
+    /**
+     * This method returns the pointer to the parent otInstance structure.
+     *
+     * @returns The pointer to the parent otInstance structure.
```
Document that this can return `NULL`.

jwhui commented on src/core/common/locator.hpp:
```diff
@@ -0,0 +1,341 @@
+    /**
+     * This method returns the pointer to the parent otInstance structure.
+     *
+     * @returns The pointer to the parent otInstance structure.
```
Document that this can return `NULL`.

jwhui commented on src/core/common/locator.hpp:
```diff
@@ -0,0 +1,341 @@
+    /**
+     * This method returns the pointer to the parent otInstance structure.
+     *
+     * @returns The pointer to the parent otInstance structure.
```
Document that this can return `NULL`.

jwhui commented on src/core/common/locator.hpp:
```diff
@@ -0,0 +1,341 @@
+    /**
+     * This method returns the pointer to the parent otInstance structure.
+     *
+     * @returns The pointer to the parent otInstance structure.
```
Document that this can return `NULL`.
